Add a limit on total number of bytes read from storage in table scan

fgwang7w opened the pull request:
Fixes #14701 

```
== RELEASE NOTES ==

General Changes
* Add `query.max-scan-physical-bytes` configuration and `query_max_scan_physical_bytes` session properties to limit total number of bytes read from storage during table scan. The default limit is 1PB.
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -312,4 +326,22 @@ private void enforceCpuLimits()
+    private void enforceScanLimits()
+    {
+        for (QueryExecution query : queryTracker.getAllQueries()) {
+            DataSize limitOpt = Optional.ofNullable(getQueryMaxScanPhysicalBytes(query.getSession()))
```
`getQueryMaxScanPhysicalBytes(session)` returns a value that can be used as is without further checking.  

This code ensures that this property defaults to what's specified in config:

```
                dataSizeProperty(
                        QUERY_MAX_SCAN_PHYSICAL_BYTES,
                        "Maximum scan physical bytes of a query",
                        queryManagerConfig.getQueryMaxScanPhysicalBytes(),
                        false),
```

and this code defines the default value that is used if nothing is specified in config file:

```
private DataSize queryMaxScanPhysicalBytes = DataSize.succinctDataSize(1, PETABYTE);
```

Hence, the code can be simplified like this:

```
        for (QueryExecution query : queryTracker.getAllQueries()) {
            DataSize limit = getQueryMaxScanPhysicalBytes(query.getSession());
            DataSize scan = query.getQueryInfo().getQueryStats().getRawInputDataSize();
            if (scan.compareTo(limit) >= 0) {
                query.fail(new ExceededScanLimitException(limit));
            }
        }
```

fgwang7w replied:
Many thanks for the proposal. I have made another version of how to implement this limit based on your suggestion. The new version is now includes a method defines in QueryExecution to collect scanned data size which is implemented in SqlQueryExecution to collect finalQueryInfo's rawInputDataSize. Please help review again.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/ExceededScanLimitException.java:
```diff
@@ -0,0 +1,28 @@
+{
+    public ExceededScanLimitException(DataSize limit)
+    {
+        super(EXCEEDED_SCAN_LIMIT, "Exceeded Scan limit of " + limit.toString());
```
Let's clarify what does the limit applies to? E.g. whether it applies to amount of data read from storage before or after compression?

fgwang7w replied:
This is the amount of consumed input bytes read from storage via QueryInfo's getQueryStats().getRawInputDataSize()

mayankgarg1990 replied:
Concurring @mbasmanova's comment - we should be clear about what scan limit has been exceeded. I will recommend `EXCEEDED_SCAN_RAW_BYTES_READ_LIMIT`. The message can also be made more explicit

fgwang7w replied:
the message has been fixed to indicate the scan bytes limit has exceeded for this exception

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -423,6 +425,11 @@ public SystemSessionProperties(
                         false),
+                dataSizeProperty(
+                        QUERY_MAX_SCAN_PHYSICAL_BYTES,
+                        "Maximum scan physical bytes of a query",
```
Let's clarify. Is this before or after compression? Before or after filtering? I assume it is after compression but before filtering.

fgwang7w replied:
Thank you Maria, this is a good point. So to clarify the confusion how the scan size is measured, I have implicitly implemented the SqlQueryExecution.getScannedBytes method to extract the query scanned size from QueryInfo's rawInputDataSize stats which is the consumed input bytes, and yes it should be the actual data size before filtering.

mayankgarg1990 replied:
Should we just say `QUERY_MAX_SCAN_RAW_INPUT_BYTES` that way it matches the public stat that we publish and it is easy for people to understand which metric did they exceed? @mbasmanova thoughts?

viczhang861 commented on presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java:
```diff
@@ -66,6 +69,8 @@
     private Duration queryMaxExecutionTime = new Duration(100, TimeUnit.DAYS);
     private Duration queryMaxCpuTime = new Duration(1_000_000_000, TimeUnit.DAYS);
 
+    private DataSize queryMaxScanPhysicalBytes = DataSize.succinctDataSize(1, PETABYTE);
```
Is `QueryStats::rawInputDataSize` equivalent to scanPhysicalBytes?  does reading from materialized intermediate table included? cc @arhimondr

mayankgarg1990 replied:
Same comment as in `SystemSessionProperties` - lets use the same name as the actual metric - `queryMaxRawInputBytes`. In accordance with that lets change the config name as well - `query.max-raw-input-bytes`.

viczhang861 commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -70,6 +74,8 @@
 
     private final Duration maxQueryCpuTime;
 
+    private final Optional<DataSize> maxQueryScanPhysicalBytes;
```
Remember to remove this.

fgwang7w replied:
consider it done, thanks

mbasmanova commented:
@fgwang7w Would you squash all 5 commits into one?

mayankgarg1990 commented on presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:
```diff
@@ -752,6 +753,23 @@ public void setUpdateType(String updateType)
+    public long getScannedBytes(StageInfo rootStage)
+    {
+        long rawInputDataSize = 0;
+        for (StageInfo stageInfo : getAllStages(Optional.of(rootStage))) {
+            if (stageInfo.getPlan().isPresent()) {
+                PlanFragment plan = stageInfo.getPlan().get();
+                if (!plan.getTableScanSchedulingOrder().isEmpty()) {
+                    Optional<QueryInfo> queryInfo = finalQueryInfo.get();
+                    if (queryInfo.isPresent()) {
+                        QueryStats queryStats = queryInfo.get().getQueryStats();
+                        rawInputDataSize += queryStats.getRawInputDataSize().toBytes();
+                    }
+                }
+            }
+        }
```
I can see a lot of issues with this logic:

1. We are reading `finalQueryInfo` which is published only when the query is finished - can you test and see if this number is actually published when the query is running ?
2. For every stage that is a scan stage, we take the whole query's bytes read and add it. So if there are 2 scan stages - in that case, you will just return `2 * bytesreadbyquery`.
3. `getAllStages` does a DFS traversal every time - and given that it will be called in a loop, this might be an expensive operation and we don't really care about the ordering here.

In my opinion - we should do something similar to the existing logic that exists for `getTotalCpuTime`. That will help ensure that the logics are similar and it is easy to change them together if we ever decide to head that path.

fgwang7w replied:
ok so this is newly implemented, the original proposal was a simple approach which is to obtain the scanned data size from `queryInfo::queryStats` if the final resolution is to do something similiar to the existing logic with `getTotalCpuTime`.
`DataSize scan = query.getQueryInfo().getQueryStats().getRawInputDataSize();`
@mbasmanova what's your suggestion on this one?

mayankgarg1990 replied:
`query.getQueryInfo().getQueryStats()` is a sort of expensive method since it aggregates all the stats and not just bytes read. We should keep this data collection as light as possible in my opinion.

tdcmeehan replied:
Agree with @mayankgarg1990 -- this can be made to be less expensive and consistent with how we calculate CPU time.

fgwang7w replied:
thank you @mayankgarg1990 @tdcmeehan for the comment, will revise this code accordingly to align with general data collection logic

tdcmeehan replied:
Were these comments addressed?

fgwang7w replied:
yes this code block has been removed as to make it simple and similar to how cpu limits enforcement is handled

mayankgarg1990 commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -312,4 +322,19 @@ private void enforceCpuLimits()
+    {
+        for (QueryExecution query : queryTracker.getAllQueries()) {
+            DataSize limit = getQueryMaxScanPhysicalBytes(query.getSession());
+            // DataSize scan = query.getQueryInfo().getQueryStats().getRawInputDataSize();
```
nit: remove

fgwang7w replied:
done

mayankgarg1990 commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -312,4 +322,19 @@ private void enforceCpuLimits()
+    private void enforceScanLimits()
+    {
+        for (QueryExecution query : queryTracker.getAllQueries()) {
+            DataSize limit = getQueryMaxScanPhysicalBytes(query.getSession());
```
Over here we are just using the session property - which a user can override. Similar to how this is being done for `enforceCpuLimits` where in we take the min of the cluster setting and session property and enforce based on that.

fgwang7w replied:
right that's the idea, getQueryMaxScanPhysicalBytes obtains the value set from system property so that it can be compared with the actual scanned value

mayankgarg1990 replied:
Properties defined in `SystemSessionProperties` are properties that can be overridden by users by saying:

```
SET SESSION query_max_scan_physical_bytes=100000000000000000;
SELECT * FROM very_large_table;
```

If as a system administrator I want to cap the amountt of data a query can read - then this doesn't work since any user can override it.

If you check `enforceCpuLimit`, there we do:

```
Duration sessionLimit = getQueryMaxCpuTime(query.getSession()); // value from SystemSessionProperties
Duration limit = Ordering.natural().min(maxQueryCpuTime, sessionLimit);  // maxQueryCpuTime is from QueryManagerConfig
```

And hence in this example - a user can lower their allowed limit, but not go beyond what the system administrator has set as the cluster setting.

fgwang7w commented:
> @fgwang7w Would you squash all 5 commits into one?

done, squashed into 1 commit now

fgwang7w commented on presto-main/src/main/java/com/facebook/presto/execution/QueryExecution.java:
```diff
@@ -53,6 +53,11 @@
 
     QueryInfo getQueryInfo();
 
+    default DataSize getScannedBytes()
+    {
+        return new DataSize(0, DataSize.Unit.BYTE);
+    }
+
```
The thought here is that we are implementing limit for scanned bytes from raw input's perspective, eventually we can add more knobs to control the scanned bytes for output data size, post-compressed size, etc. the default scannedBytes gives a general guideline for all implementations with datasize=0b to start with

mayankgarg1990 replied:
My apologies - your point is not very clear to me. `scannedBytes for output data size` - should we just not call it `getOutputDataBytes`. I am not sure what kind of guidance are we offering when we start with `datasize=0b`

tdcmeehan replied:
I was reading through this again and have the same question.  Why make this a default, why not just move this default method to `DataDefinitionExecution`?

fgwang7w replied:
Sorry, missed this comment. yes agree with you all, so instead of going to default value, I will just remove this definition and simply take session value and config value into account. If we want to implement further based on scannedbyte config like scan limit for different connectors, we can add it back to set a bar as a default scan byte limit for general purpose, but for now I will remove this block of code

mbasmanova commented:
@fgwang7w I'm in meetings all day today and won't have time to review/answer questions on this PR before the long weekend. I'll take a look early next week.

fgwang7w commented:
> @fgwang7w I'm in meetings all day today and won't have time to review/answer questions on this PR before the long weekend. I'll take a look early next week.

Sure thank you, and in the meantime I will do more testing to ensure quality of the code

fgwang7w commented:
hi @mbasmanova please help review the revised commit, many thanks~

mbasmanova commented:
@fgwang7w Looks good to, but I'll defer to @mayankgarg1990 and @tdcmeehan to confirm that their comments have been addressed properly. Commit message needs to be updated to match the guidelines at https://chris.beams.io/posts/git-commit/ . For example, 

```
Add a limit on total number of bytes read from storage in table scan

Add query.max-scan-physical-bytes configuration and query_max_scan_physical_bytes 
session properties to limit the total number of bytes reads from storage during table scan. 
The default limit is 1PB.
```

mbasmanova commented:
@mayankgarg1990 @tdcmeehan Mayank, Tim, would you take another look?

mayankgarg1990 commented:
@mbasmanova , this is on my radar, I will get to it by tomorrow (7/22) :)

mbasmanova commented:
Thank you, Mayank.

fgwang7w commented:
@mayankgarg1990 Hi Could you please help review the commit for the upcoming release merge? many thanks!

mayankgarg1990 commented:
@fgwang7w - As @tdcmeehan pointed out - my comments from my previous review are not addressed yet. @tdcmeehan  commented 2 days ago

fgwang7w commented:
> @fgwang7w - As @tdcmeehan pointed out - my comments from my previous review are not addressed yet. @tdcmeehan commented 2 days ago

yes I resubmitted the squashed commit code just now and replied to both you @mayankgarg1990  and @tdcmeehan regarding the querystats perf issue for getScannedBytes method and default scan limit. Basically I have simplified the code and reduce unnecessary method calls per suggestions. The default scan limit is also removed for now. Please give another round of review, many thanks!

mayankgarg1990 commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -312,4 +324,21 @@ private void enforceCpuLimits()
+    private void enforceScanLimits()
+    {
+        for (QueryExecution query : queryTracker.getAllQueries()) {
+            BasicQueryStats basicQueryStats = query.getBasicQueryInfo().getQueryStats();
+            QueryStats queryStats = query.getQueryInfo().getQueryStats();
```
these 2 are not being used - remove these

fgwang7w replied:
sure, it's removed

mayankgarg1990 commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -312,4 +324,21 @@ private void enforceCpuLimits()
+        for (QueryExecution query : queryTracker.getAllQueries()) {
+            BasicQueryStats basicQueryStats = query.getBasicQueryInfo().getQueryStats();
+            QueryStats queryStats = query.getQueryInfo().getQueryStats();
+            DataSize scan = query.getBasicQueryInfo().getQueryStats().getRawInputDataSize();
```
This is still a bit expensive since getting querystats involves -> for all stages, for all tasks, for all metrics.

The way CPU does it is still cheaper -> all stages, all tasks, just the cpu - so we can do something similar here and make it even cheaper - 

We can follow a similar flow here

SqlQueryManager -> SqlQuerySchedulerInterface#getTotalCpuTime -> SqlStageExecution#getTotalCpuTime

and just sum up the bytes involved.

fgwang7w replied:
I'm not certain if your approach about summing up bytes from `SqlStageExecution#getTotalScanBytes` is accurate... it's always the best practice to inherit scanned byte size from existing querystats to ensure the result is correct. I suggest we maintain current implementation with `SqlQueryManager:getBasicQueryInfo-> BasicQueryStats:getQueryStats` . How much cheaper are we suggesting to reduce if we want to bypass the existing logic with a risk that we might have a false result? 
My option is that `rawInputDataSize` should be the only reliable source of truth when setting against with the hard scan limit.

mayankgarg1990 replied:
I don't agree that `BasicQueryStats:getQueryStats` is the only trusted source. As you can see, we are already doing this for total cpu ms and memory limits so I don't see why raw input bytes will be any different here. Again, my only concern is that this is a single thread already doing cpu and memory enforcements and we should keep it as light weight as possible and by keeping this trend, we will ensure that the new entries that are added also follow this lighter weight approach.

fgwang7w replied:
understood, thank you! I have implemented your approach, I think it is safer and cheaper to acquire actual number via `SqlQuerySchedulerInterface::getRawInputDataSize` -> `QueryExecution::getRawInputDataSize`, and sum up all task's stats via `SqlStageExecution::getRawInputDataSize`. Please review the revised commit again, many thanks!

fgwang7w commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -424,6 +426,11 @@ public SystemSessionProperties(
                         false,
                         false),
+                dataSizeProperty(
+                        QUERY_MAX_SCAN_PHYSICAL_BYTES,
```
sure actually I would synchronize all method call names and varilables to QueryMaxScanRawInputBytes to map with rawInputBytes

mayankgarg1990 commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -312,4 +323,19 @@ private void enforceCpuLimits()
+    private void enforceScanLimits()
+    {
+        for (QueryExecution query : queryTracker.getAllQueries()) {
+            DataSize scan = query.getRawInputDataSize();
```
nit - lets rename this as `rawInputSize`

fgwang7w replied:
fixed, thanks

mayankgarg1990 commented on presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java:
```diff
@@ -333,6 +334,14 @@ public synchronized Duration getTotalCpuTime()
     }
 
+    public synchronized DataSize getRawInputDataSize()
+    {
+        long datasize = getAllTasks().stream()
```
Lets add this check at the top to ensure that we are only considering table scan nodes - 

```
if (planFragment.getTableScanSchedulingOrder().isEmpty()) {
    return new DataSize(0, BYTE);
}
```

fgwang7w replied:
agree, we need to bypass when source is empty, fixed

mayankgarg1990 commented on presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java:
```diff
@@ -66,6 +69,8 @@
     private Duration queryMaxExecutionTime = new Duration(100, TimeUnit.DAYS);
     private Duration queryMaxCpuTime = new Duration(1_000_000_000, TimeUnit.DAYS);
 
+    private DataSize queryMaxScanRawInputBytes = DataSize.succinctDataSize(1, PETABYTE);
```
Lets set the default to a higher number to avoid unexpected failures when people deploy this new version. Every sys admin should be able to set a reasonable value in their configurations. How about an exabyte (1000, PETABYTE)?

fgwang7w replied:
sure, this field is adjustable anytime by DBAs. Default is revised to 1000PB.

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/session/PropertyMetadata.java:
```diff
@@ -199,4 +200,17 @@ public Object encode(T value)
                 object -> object);
     }
+
+    public static PropertyMetadata<DataSize> dataSizeProperty(String name, String description, DataSize defaultValue, boolean hidden)
```
Nice
